optaplanner-core/src/test/java/org/optaplanner/core/config/solver/SolverConfigMultiThreadedTest.java

```diff
@@ -0,0 +1,131 @@
+package org.optaplanner.core.config.solver;
+
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+import org.junit.Ignore;
+import org.junit.Test;
+import org.optaplanner.core.api.solver.Solver;
+import org.optaplanner.core.api.solver.SolverFactory;
+import org.optaplanner.core.config.solver.testutil.TestThreadFactory;
+import org.optaplanner.core.impl.testdata.domain.TestdataEntity;
+import org.optaplanner.core.impl.testdata.domain.TestdataSolution;
+import org.optaplanner.core.impl.testdata.domain.TestdataValue;
+import org.optaplanner.core.impl.testdata.util.PlannerTestUtils;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertNull;
+import static org.junit.Assert.assertSame;
+import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.fail;
+import static org.mockito.Mockito.spy;
+import static org.mockito.Mockito.when;
+
+public class SolverConfigMultiThreadedTest {
+
+    @Test
+    public void moveThreadCountAutoIsCorrectlyResolvedWhenCpuCountIsPositive() {
+        final int cpuCount = 16;
+        assertEquals(Integer.valueOf(cpuCount - 2), mockSolverConfigForMoveThreadCountAuto(cpuCount).resolveMoveThreadCount());
+    }
+
+    @Test
+    public void moveThreadCountAutoIsResolvedToNullWhenCpuCountIsNegative() {
+        final int cpuCount = -2;
+        assertNull(mockSolverConfigForMoveThreadCountAuto(cpuCount).resolveMoveThreadCount());
+    }
+
+    private SolverConfig mockSolverConfigForMoveThreadCountAuto(int mockCpuCount) {
+        SolverConfig solverConfig = spy(SolverConfig.class);
+        when(solverConfig.getAvailableProcessors()).thenReturn(mockCpuCount);
+        solverConfig.setMoveThreadCount(SolverConfig.MOVE_THREAD_COUNT_AUTO);
+        return solverConfig;
+    }
+
+    @Test
+    public void moveThreadCountIsCorrectlyResolvedWhenValueIsPositive() {
+        SolverConfig solverConfig = new SolverConfig();
+        solverConfig.setMoveThreadCount("2");
+        assertEquals(Integer.valueOf(2), solverConfig.resolveMoveThreadCount());
+    }
+
+    @Test
+    public void moveThreadCountThrowsExceptionWhenValueIsNegative() {
+        SolverConfig solverConfig = new SolverConfig();
+        solverConfig.setMoveThreadCount("-6");
+        try {
+            solverConfig.resolveMoveThreadCount();
+            fail("IllegalArgumentException should have been thrown.");
+        } catch (IllegalArgumentException expectedException) {
+            //expected
```

Comment:
Is this possible? 
@Test(expectedException = IllegalArgumentException.class)

Reply:
I think that, even though it is possible, it's the old style of checking exceptions. The way Radek wrote it, although it requires some boilerplate code, it is clearly visible which statement throws the exception, it's possible to check exception message and even continue with the test after the exception is thrown.

Once we migrate to JUnit 5, it will be possible to put it in a more elegant way:
```java


    @Test
    void exceptionTesting() {
        Throwable exception = assertThrows(IllegalArgumentException.class, () -> {
            solverConfig.resolveMoveThreadCount();
        });
        assertThat(exception).hasMessageContaining("negative");
    }
```
and Radek's style will make the migration easier.

Reply:
@yurloc Geoffrey's motivation for this change is to make searching for all occurrences of error checking easier - in case we have only @Test(expectedException=...), there is 0 chance to miss it.
I have checked the current source code and found out that we already mix both style - now it doesn't make sense to change it here. I will keep the test as is.